### PR TITLE
websocket test refactor with Junit and arquillian

### PR DESCRIPTION
**Pull request by gurunrao**
Websocket Platform TCK and Standalone TCK  test refactor using Junit and arquillian.

Standalone Test results:

> The CI run for the refactored can be found at: https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-tck-glassfish/20/consoleFull
> 
> The test logs can be found at: https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-tck-glassfish/20/artifact/glassfish-runner/websocket-tck/target/failsafe-reports/
> 
> Glassfish logs at: [https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-tck-glassfish/20/artifact/glassfish-runner/websocket-tck/target/glassfish7/glassfish/domains/domain1/logs/server.log](https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-tck-glassfish/20/artifact/glassfish-runner/websocket-tck/target/glassfish7/glassfish/domains/domain1/logs/server.log)
> 
>Test run summary:
> Tests run: 736, Failures: 0, Errors: 0, Skipped: 0

Platform Test results:

> The CI run for the refactored can be found at: https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-platform-tck-glassfish/2/consoleFull
> 
> The test logs can be found at: https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-platform-tck-glassfish/2/artifact/glassfish-runner/websocket-platform-tck/target/failsafe-reports/
> 
> Glassfish logs at: [https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-tck-glassfish/2/artifact/glassfish-runner/websocket-platform-tck/target/glassfish7/glassfish/domains/domain1/logs/server.log](https://ci.eclipse.org/jakartaee-tck/job/guru/job/jakarta-websocket-platform-tck-glassfish/2/artifact/glassfish-runner/websocket-tck/target/glassfish7/glassfish/domains/domain1/logs/server.log)
> 
> Test run summary:
> Tests run: 748, Failures: 0, Errors: 0, Skipped: 0

### CODE REVIEW

**Comment by olamy** [comment]
is it using arquilian?

**Comment by gurunrao** [comment]
> is it using arquilian?

Yes, it uses arquillian.

**Comment by olamy** [comment]
> > is it using arquilian?
> 
> Yes, it uses arquillian and junit5.

great. I will have a try next week with Jetty.

**Comment by gurunrao** [comment]
Discussion thread for negdep failures - https://www.eclipse.org/lists/websocket-dev/msg00184.html

**Comment by olamy** [comment]
Please have a look at this PR https://github.com/gurunrao/jakartaee-tck/pull/8
The host:port should be coming from Arquillian. Some container such jetty just start some random so we cannot expect to have hardcoded port before starting the build

File: websocket/src/main/java/com/sun/ts/tests/websocket/api/jakarta/websocket/clientendpointconfig/WSClientIT.java [hunk 1]
```diff
@@ -0,0 +1,1094 @@
...
+		if (passed == false) {
+			throw new Exception("Test failed: " + log.toString());
+		}
+	}
+
+	public void cleanup() {
+	}
+}
+
+class TCKConfigurator extends Configurator {
```
Comment (olamy) [inline websocket/src/main/java/com/sun/ts/tests/websocket/api/jakarta/websocket/clientendpointconfig/WSClientIT.java:1085]: this  looks un common syntax for definition of a class. Should it be in a separate file or just a static internal class.
Remove `}` from line 1083 and move it at the end

**Review by olamy — CHANGES_REQUESTED** [review]

**Comment by olamy** [comment]
why renaming all classes to `*IT`?

**Comment by gurunrao** [comment]
> why renaming all classes to `*IT`?

fail-safe plugin default inclusion wildcards are as follows:
https://maven.apache.org/surefire/maven-failsafe-plugin/examples/inclusion-exclusion.html

File: websocket/src/main/java/com/sun/ts/tests/websocket/api/jakarta/websocket/clientendpointconfig/WSClientIT.java (hunk 1 above)
Comment (gurunrao) [reply websocket/src/main/java/com/sun/ts/tests/websocket/api/jakarta/websocket/clientendpointconfig/WSClientIT.java:1085]: code is same as https://github.com/jakartaee/platform-tck/blob/master/src/com/sun/ts/tests/websocket/api/jakarta/websocket/clientendpointconfig/WSClient.java#L1158

**Comment by gurunrao** [comment]
> Please have a look at this PR [gurunrao#8](https://github.com/gurunrao/jakartaee-tck/pull/8) The host:port should be coming from Arquillian. Some container such jetty just start some random so we cannot expect to have hardcoded port before starting the build

The host:port were configured [before](https://github.com/jakartaee/platform-tck/blob/master/install/websocket/bin/ts.jte#L177), we intend to keep the functionality as before.

**Comment by olamy** [comment]
> > Please have a look at this PR [gurunrao#8](https://github.com/gurunrao/jakartaee-tck/pull/8) The host:port should be coming from Arquillian. Some container such jetty just start some random so we cannot expect to have hardcoded port before starting the build
> 
> 
> 
> The host:port were configured in jtest and harness TCK (https://github.com/jakartaee/platform-tck/blob/master/install/websocket/bin/ts.jte#L177), we intend to keep the functionality as before. 

Yes but now with Arquillian this works totally different and the change from the PR must be applied. I have tested with gf runner and this works. This has been done as this with servlet tck.
If not changed there jetty or any arquillian container using random port will not being able to use this TCK code

**Comment by joakime** [comment]
> The host:port were configured in jtest and harness TCK (https://github.com/jakartaee/platform-tck/blob/master/install/websocket/bin/ts.jte#L177), we intend to keep the functionality as before.

I don't like this.
The hardcoded port should not be the default or mandatory behavior.
All of the bad behaviors of jtest should be considered deprecated legacy behaviors.
Use junit and arquillian properly and don't hang onto the bad decisions from the past.

**Comment by gurunrao** [comment]
> > The host:port were configured in jtest and harness TCK (https://github.com/jakartaee/platform-tck/blob/master/install/websocket/bin/ts.jte#L177), we intend to keep the functionality as before.
> 
> I don't like this. The hardcoded port should not be the default or mandatory behavior. All of the bad behaviors of jtest should be considered deprecated legacy behaviors. Use junit and arquillian properly and don't hang onto the bad decisions from the past.

Port is configured using properties file and not harcoded.

**Comment by joakime** [comment]
> [gurunrao#8](https://github.com/gurunrao/jakartaee-tck/pull/8) - doesn't take care of secure ports

The example of gurunrao#8 shows how it works with a default deployment.

``` java
	@ArquillianResource
	@OperateOnDeployment("_DEFAULT_")
	public URL url;
```

Take the Servlet world, for example, if the need is for an Arquillian operation with a secure port, then it is accessed with a operational name.

https://github.com/jakartaee/platform-tck/blob/5d9911b1755fecf66234f162600793969de3bd8e/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcertanno/Client.java#L63-L64

and used like ...

https://github.com/jakartaee/platform-tck/blob/5d9911b1755fecf66234f162600793969de3bd8e/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcertanno/Client.java#L112-L113

**Comment by markt-asf** [comment]
Some feedback from testing this with Tomcat.

First the big plus in the improved speed. Tests execute much faster with the entire suite complete in just over 3 minutes.

It took a little while to figure out how to configure the TCK to run with Tomcat but now I have figured that it it will be easy to maintain going forward and to potentially feed into CI systems to run the TCK more regularly.

I am currently seeing a small number (three) of test failures. I need to look into these to figure out if it is my configuration of the test suite that is still not quite right or if something else is going on.

**Comment by markt-asf** [comment]
There does appear to be a concurrency issue with the refactored tests.

A number of the tests define `private static StringBuffer receivedMessageString` which is then used to log messages at key points thoughout the test which are then validated at the end of the test. The Maven Failsafe plugin runs all the tests in a single JVM by default and that appeared to be causing issues with the validation of `receivedMessageString`.

I'm not sure what the "right" fix is but configuring the tests to run with `<reuseForks>false</reuseForks>` fixed those failures. The downside is that the tests run in just over 5 minutes rather than just over 3 minutes (which is still much faster than the old TCK). The "right" fix might be breaking up those tests into separate classes each with their own `private static StringBuffer receivedMessageString` so they don't interfere.

I'm still seeing a couple of timeout tests fail but at the moment I think this is a Tomcat configuration issue.

**Comment by markt-asf** [comment]
It isn't a Tomcat issue. The root cause is the same as the other failures. The problem is that `setTimeout2Test` sees the message `TCKTestServer second message after sleep` generated by `setTimeout1Test` and fails. My quick fix was to use unique messages for the two timeout tests.

That has resolved the failures I was seeing but it looks like one of the timeout tests is now failing further on. I'm still investigating.

**Comment by markt-asf** [comment]
Woot! The remaining issues were configuration (I'd forgotten a system property needed to ensure the timeouts fire promptly enough for the tests) and a Tomact bug caused by a recent refactoring. All 736 tests are now passing which is consistent with the old TCK tests that had 737 tests (including the signature test which hasn't been refactored yet).

**Comment by olamy** [comment]
@gurunrao please read my comment here https://github.com/jakartaee/platform-tck/pull/1182#discussion_r1243639759
maybe we can have something which could make everybody happy?

**Comment by gurunrao** [comment]
> > [gurunrao#8](https://github.com/gurunrao/jakartaee-tck/pull/8) - doesn't take care of secure ports
> 
> The example of [gurunrao#8](https://github.com/gurunrao/jakartaee-tck/pull/8) shows how it works with a default deployment.
> 
> ```java
> 	@ArquillianResource
> 	@OperateOnDeployment("_DEFAULT_")
> 	public URL url;
> ```
> 
> Take the Servlet world, for example, if the need is for an Arquillian operation with a secure port, then it is accessed with a operational name.
> 
> https://github.com/jakartaee/platform-tck/blob/5d9911b1755fecf66234f162600793969de3bd8e/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcertanno/Client.java#L63-L64
> 
> and used like ...
> 
> https://github.com/jakartaee/platform-tck/blob/5d9911b1755fecf66234f162600793969de3bd8e/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcertanno/Client.java#L112-L113
> 
> Which would be needed in only one websocket test class ... websocket/src/main/java/com/sun/ts/tests/websocket/platform/jakarta/websocket/server/handshakerequest/authenticatedssl/WSCClientIT.java

I have made necessary changes as per comments, the ssl port auto-detection is not getting detected.

**Comment by gurunrao** [comment]
> > > [gurunrao#8](https://github.com/gurunrao/jakartaee-tck/pull/8) - doesn't take care of secure ports
> > 
> > 
> > The example of [gurunrao#8](https://github.com/gurunrao/jakartaee-tck/pull/8) shows how it works with a default deployment. Please refer latest code.
> > ```java
> > 	@ArquillianResource
> > 	@OperateOnDeployment("_DEFAULT_")
> > 	public URL url;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Take the Servlet world, for example, if the need is for an Arquillian operation with a secure port, then it is accessed with a operational name.
> > https://github.com/jakartaee/platform-tck/blob/5d9911b1755fecf66234f162600793969de3bd8e/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcertanno/Client.java#L63-L64
> > 
> > and used like ...
> > https://github.com/jakartaee/platform-tck/blob/5d9911b1755fecf66234f162600793969de3bd8e/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcertanno/Client.java#L112-L113
> > 
> > Which would be needed in only one websocket test class ... websocket/src/main/java/com/sun/ts/tests/websocket/platform/jakarta/websocket/server/handshakerequest/authenticatedssl/WSCClientIT.java
> 
> I have made necessary changes as per comments, the ssl port auto-detection is not getting detected.

@olamy @joakime any suggestions on SSL port auto-detection working?

**Comment by olamy** [comment]
@gurunrao I'm on holidays. I can have a look next week.
This is how we do for Jetty https://github.com/jetty-project/servlet-tck-run/blob/jetty-12-ee10/src/test/resources/arquillian.xml

as you can see there is the default 
```
<container qualifier="http" default="true">
```
and then the `https` one 
```
<container qualifier="https">
```

`https` here it's just an id in the arquillian configuration.

**Comment by scottmarlow** [comment]
We should consider asking on Websocket dev mailing list if the Standalone WebSocket TCK tests should move to the WebSocket spec.  I see that Mark Thomas is here helping already but its not too early to discuss moving the WebSocket TCK work to the Specification.  

We started that conversation recently on the [Servlet mailing list](https://www.eclipse.org/lists/servlet-dev/threads.html#00580).

**Comment by gurunrao** [comment]
> @gurunrao I'm on holidays. I can have a look next week. This is how we do for Jetty https://github.com/jetty-project/servlet-tck-run/blob/jetty-12-ee10/src/test/resources/arquillian.xml
> 
> as you can see there is the default
> 
> ```
> <container qualifier="http" default="true">
> ```
> 
> and then the `https` one
> 
> ```
> <container qualifier="https">
> ```
> 
> `https` here it's just an id in the arquillian configuration.
> 
> And you reference this as is in your test.
> 
> https://github.com/jakartaee/platform-tck/blob/644743ff162e0a322601b4c9d939f4a517cc2316/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcert/Client.java#L73

https://github.com/gurunrao/jakartaee-tck/blob/tck-refactor-websocket-2/glassfish-runner/websocket-platform-tck/src/test/resources/arquillian.xml#L10 has details on container.

**Comment by gurunrao** [comment]
I am planning to merge the PR, if there are no more review comments.

**Comment by gurunrao** [comment]
> > @gurunrao I'm on holidays. I can have a look next week. This is how we do for Jetty https://github.com/jetty-project/servlet-tck-run/blob/jetty-12-ee10/src/test/resources/arquillian.xml
> > as you can see there is the default
> > ```
> > <container qualifier="http" default="true">
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and then the `https` one
> > ```
> > <container qualifier="https">
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `https` here it's just an id in the arquillian configuration.
> > And you reference this as is in your test.
> > https://github.com/jakartaee/platform-tck/blob/644743ff162e0a322601b4c9d939f4a517cc2316/servlet/src/main/java/com/sun/ts/tests/servlet/spec/security/clientcert/Client.java#L73
> 
> https://github.com/gurunrao/jakartaee-tck/blob/tck-refactor-websocket-2/glassfish-runner/websocket-platform-tck/src/test/resources/arquillian.xml#L10 has details on container.

@olamy - As you suggested SSL port auto-detection might be issue with glassfish arquillian, since we have system properties, I will not debug the auto detection of SSL further.

**Pull request merged**